src/libraries/System.Net.Security/src/System/Net/Security/Pal.OSX/SafeDeleteNwContext.cs

```diff
@@ -0,0 +1,574 @@
+
+        // Handshake state machine
+        private enum HandshakeState
+        {
+            NotStarted,
+            InProgress,
+            WaitingForPeer,
+            Completed,
+            Failed
+        }
+
+        private readonly Stream _transportStream;
+        private readonly SslAuthenticationOptions _sslAuthenticationOptions;
+        private TaskCompletionSource<Exception?> _handshakeCompletionSource = new TaskCompletionSource<Exception?>(TaskCreationOptions.RunContinuationsAsynchronously);
+        private Task? _transportReadTask;
+        private ArrayBuffer _appReceiveBuffer = new(InitialBufferSize);
+        private CancellationTokenSource _shutdownCts = new CancellationTokenSource();
+
+        // Buffers
+        private const int InitialBufferSize = 2 * 1024;
+        private ArrayBuffer _outputBuffer = new(InitialBufferSize);
+
+        private bool _disposed;
+        private bool _clientCertificateRequested;
+
+        public SafeDeleteNwContext(SslAuthenticationOptions sslAuthenticationOptions, Stream transportStream) : base(IntPtr.Zero)
+        {
+            _connectionHandle = Interop.NetworkFramework.Tls.CreateContext(sslAuthenticationOptions.IsServer);
+            _transportStream = transportStream;
+            _sslAuthenticationOptions = sslAuthenticationOptions;
+            if (_connectionHandle.IsInvalid)
+            {
+                throw new Exception("Failed to create Network Framework connection"); // TODO: Make this string resource
+            }
+
+            ValidateSslAuthenticationOptions(sslAuthenticationOptions);
+            _thisHandle = GCHandle.Alloc(this, GCHandleType.Normal);
+            SetTlsOptions(sslAuthenticationOptions);
+        }
+
+        public static bool IsNetworkFrameworkAvailable => IsSwitchEnabled && s_isNetworkFrameworkAvailable.Value;
+
+        internal bool ClientCertificateRequested => _clientCertificateRequested;
+
+        internal Task<Exception?> HandshakeAsync(CancellationToken cancellationToken)
+        {
+            Interop.NetworkFramework.Tls.StartTlsHandshake(_connectionHandle, GCHandle.ToIntPtr(_thisHandle));
+
+            // TODO: Handle cancellation token
+
+            _transportReadTask = Task.Run(async () =>
+            {
+                try
+                {
+                    byte[] buffer = new byte[16 * 1024];
+                    Memory<byte> readBuffer = new Memory<byte>(buffer);
+
+                    while (!_shutdownCts.IsCancellationRequested)
+                    {
+                        // Read data from the transport stream
+                        int bytesRead = await _transportStream.ReadAsync(readBuffer, _shutdownCts.Token).ConfigureAwait(false);
+
+                        if (bytesRead > 0)
+                        {
+                            // Process the read data
+                            await WriteInboundWireDataAsync(readBuffer.Slice(0, bytesRead)).ConfigureAwait(false);
```

Comment:
I'm wondering if we could abstract this inside `ProcessTlsFrame` and call SslStreamPal.ProcessTlsFrame(). We sort  of do already but there is also more logging and processing.